File: arrays-Hashing/leetCode_242.py
"""
Given two strings s and t, return true if t is an 
anagram of s, and false otherwise.
"""
class Solution:
    def isAnagram(self, s, t):
        if len(s) != len(t):
            return False
        s_dict = {}
        t_dict = {}
        for i in range(len(s)):
            if s[i] in s_dict:
                s_dict[s[i]] += 1
            else:
                s_dict[s[i]] = 1
            if t[i] in t_dict:
                t_dict[t[i]] += 1
            else:
                t_dict[t[i]] = 1
        return s_dict == t_dict

s = Solution()
print(s.isAnagram("anagram", "nagaram"))
print(s.isAnagram("rat", "car"))

File: arrays-Hashing/leetCode_242.java
import java.util.HashMap;
import java.util.Map;

public class leetCode_242 {
    public boolean isAnagram(String s, String t){
        Map<Character, Integer> map = new HashMap<>();
        for (char c : s.toCharArray()){
            map.put(c, map.getOrDefault(c, 0) + 1);
        }
        for (char x : t.toCharArray()){
            map.put(x, map.getOrDefault(x, 0) - 1);
        }
        for (int val : map.values()){
            if (val != 0){
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args){
        leetCode_242 solution = new leetCode_242();
        String s = "anagram";
        String t = "nagaram";
        boolean result = solution.isAnagram(s, t);
        System.out.println("Is anagram: " + result);
    }
}
